File: api/gateway/native/src/routes/v2/book.ts
import express, { NextFunction, Request, Response } from 'express'
import { getBook, getBookByIsbn, listBookReview, listUserWithUserIds } from '~/api'
import { IGetBookByIsbnInput, IGetBookInput, IListBookReviewInput, IListUserByUserIdsInput } from '~/types/input'
import {
  IBookOutput,
  IBookOutputAuthor,
  IReviewListOutput,
  IReviewListOutputReview,
  IUserHashOutput,
} from '~/types/output'
import { IBookV2Response, IBookV2ResponseReview, IBookV2ResponseUser } from '~/types/response'
import { LIST_DEFAULT_LIMIT, LIST_DEFAULT_OFFSET, LIST_DEFAULT_ORDER_BY, LIST_DEFAULT_ORDER_DIRECTION } from '~/util'

const router = express.Router()

router.get(
  '/v2/books/:bookId',
  async (req: Request, res: Response<IBookV2Response>, next: NextFunction): Promise<void> => {
    const { bookId } = req.params
    const { key } = req.query as { [key: string]: string }

    await bookByIdOrIsbn(req, key, bookId)
      .then(async (bookOutput: IBookOutput) => {
        const reviewListInput: IListBookReviewInput = {
          bookId: bookOutput.id,
          limit: LIST_DEFAULT_LIMIT,
          offset: LIST_DEFAULT_OFFSET,
          by: LIST_DEFAULT_ORDER_BY,
          direction: LIST_DEFAULT_ORDER_DIRECTION,
        }

        const reviewsOutput = await listBookReview(req, reviewListInput)

        const userIds = reviewsOutput.reviews.map((rv: IReviewListOutputReview): string => {
          return rv.userId
        })
        const userListInput: IListUserByUserIdsInput = {
          ids: Array.from(new Set(userIds)),
        }

        const usersOutput = await listUserWithUserIds(req, userListInput).catch(() => {
          return {}
        })

        return setBookResponse(bookOutput, reviewsOutput, usersOutput)
      })
      .then((response: IBookV2Response) => {
        res.status(200).json(response)
      })
      .catch((err: Error) => {
        next(err)
      })
  }
)

async function bookByIdOrIsbn(req: Request, key: string, value: string): Promise<IBookOutput> {
  switch (key) {
    case 'isbn': {
      const isbn: string = value
      return bookByIsbn(req, isbn)
    }
    case 'id':
    default: {
      const id: number = Number(value) || 0
      return bookById(req, id)
    }
  }
}

async function bookById(req: Request, bookId: number): Promise<IBookOutput> {
  const input: IGetBookInput = {
    bookId,
  }

  return getBook(req, input)
}

function bookByIsbn(req: Request, isbn: string): Promise<IBookOutput> {
  const input: IGetBookByIsbnInput = {
    isbn,
  }

  return getBookByIsbn(req, input)
}

function setBookResponse(
  bookOutput: IBookOutput,
  reviewsOutput: IReviewListOutput,
  usersOutput: IUserHashOutput
): IBookV2Response {
  const authorNames = bookOutput.authors.map((item: IBookOutputAuthor): string => {
    return item.name
  })

  const authorNameKanas = bookOutput.authors.map((item: IBookOutputAuthor): string => {
    return item.nameKana
  })

  const reviews = reviewsOutput.reviews.map(
    (item: IReviewListOutputReview): IBookV2ResponseReview => {
      const user: IBookV2ResponseUser = {
        id: '',
        username: 'unknown',
        thumbnailUrl: '',
      }

      if (usersOutput[item.userId]) {
        const { id, username, thumbnailUrl } = usersOutput[item.userId]

        user.id = id
        user.username = username
        user.thumbnailUrl = thumbnailUrl
      }

      return {
        id: item.id,
        impression: item.impression,
        createdAt: item.createdAt,
        updatedAt: item.updatedAt,
        user,
      }
    }
  )

  const response: IBookV2Response = {
    id: bookOutput.id,
    title: bookOutput.title,
    titleKana: bookOutput.titleKana,
    description: bookOutput.description,
    isbn: bookOutput.isbn,
    publisher: bookOutput.publisher,
    publishedOn: bookOutput.publishedOn,
    thumbnailUrl: bookOutput.thumbnailUrl,
    rakutenUrl: bookOutput.rakutenUrl,
    size: bookOutput.rakutenSize,
    author: authorNames.join('/'),
    authorKana: authorNameKanas.join('/'),
    createdAt: bookOutput.createdAt,
    updatedAt: bookOutput.updatedAt,
    reviews,
    reviewLimit: reviewsOutput.limit,
    reviewOffset: reviewsOutput.offset,
    reviewTotal: reviewsOutput.total,
  }

  return response
}

export default router

File: api/internal/gateway/native/entity/book.go
package entity

import (
	"strings"

	"github.com/calmato/gran-book/api/internal/gateway/entity"
)

type Book struct {
	ID           int64       `json:"id"`           // 書籍ID
	Title        string      `json:"title"`        // タイトル
	TitleKana    string      `json:"titleKana"`    // タイトル(かな)
	Description  string      `json:"description"`  // 説明
	Isbn         string      `json:"isbn"`         // ISBN
	Publisher    string      `json:"publisher"`    // 出版社名
	PublishedOn  string      `json:"publishedOn"`  // 出版日
	ThumbnailURL string      `json:"thumbnailUrl"` // サムネイルURL
	RakutenURL   string      `json:"rakutenUrl"`   // 楽天ショップURL
	Size         string      `json:"size"`         // 楽天書籍サイズ
	Author       string      `json:"author"`       // 著者名一覧
	AuthorKana   string      `json:"authorKana"`   // 著者名一覧(かな)
	Reviews      BookReviews `json:"reviewsList"`  // レビュー一覧
	ReviewLimit  int64       `json:"reviewLimit"`  // レビュー取得上限
	ReviewOffset int64       `json:"reviewOffset"` // レビュー取得開始位置
	ReviewTotal  int64       `json:"reviewTotal"`  // レビュー検索一致件数
	CreatedAt    string      `json:"createdAt"`    // 登録日時
	UpdatedAt    string      `json:"updatedAt"`    // 更新日時
}

func NewBook(b *entity.Book, rs BookReviews) *Book {
	return &Book{
		ID:           b.Id,
		Title:        b.Title,
		TitleKana:    b.TitleKana,
		Description:  b.Description,
		Isbn:         b.Isbn,
		Publisher:    b.Publisher,
		PublishedOn:  b.PublishedOn,
		ThumbnailURL: b.ThumbnailUrl,
		RakutenURL:   b.RakutenUrl,
		Size:         b.RakutenSize,
		Author:       strings.Join(b.AuthorNames(), "/"),
		AuthorKana:   strings.Join(b.AuthorNameKanas(), "/"),
		CreatedAt:    b.CreatedAt,
		UpdatedAt:    b.UpdatedAt,
		Reviews:      rs,
	}
}

func (b *Book) Fill(limit, offset, total int64) {
	b.ReviewLimit = limit
	b.ReviewOffset = offset
	b.ReviewTotal = total
}
